# s03-Window Functions within SQLite

Window Functions get available sincle SQLite 3.25.According to the SQLite docs:

>A window function is a special SQL function where the input values are taken from a “window” of one or more rows in the results set of a SELECT statement.

SQLite now has these built-in Window functions:

- row_number()
- rank()
- dense_rank()
- percent_rank()
- cume_dist()
- ntile(N)
- lag(expr), lag(expr, offset), lag(expr, offset, default)
- lead(expr), lead(expr, offset), lead(expr, offset, default)
- first_value(expr)
- last_value(expr)
- nth_value(expr, N)

Previous tutorials are created based on a little bit older SQLite versions. At that time, window Functions are not available. In this tutoiral, I am going to touch some of them in the following 5 sections:

- Firstly, give a comparison between SQL Window Functions vs. SQL Aggregate Functions

- Secondly, go through a few WINDOW functions based on regular aggregate functions, such as **AVG**, **MIN/MAX**, **COUNT**, **SUM**.

- Thirdly, focus on rank-related functions, such as **ROW_NUMBER**, **RANK**,and **RANK_DENSE**. Those functions are handy when generating ranking indexes, and you need to be fluent in those functions before entering a data scientist SQL interview.

- Forthly, talk about generating statistics (e.g., percentiles, quartiles, median, etc .) with the **NTILE** function, a common task for a data scientist.

- Fifthly, focus on **LAG** and **LEAD**, two functions that are super important if you are interviewing for a role that requires dealing with time-series data.


## 1. [Window Functions vs. Aggregate Functions](https://learnsql.com/blog/window-functions-vs-aggregate-functions/)

In the previous tutorials, we presented some application cases of Aggregate Functions. What are the Similarities and Differences between them? Let's have a quick gothrough.

### 1.1 Aggregate functions

*Aggregate functions* operate on a set of values to return a single scalar value. These are SQL aggregate functions:

- AVG() returns the average of the specified values.
- SUM() calculates the sum of all values in the set.
- MAX() and MIN() return the maximum and minimum value, respectively.
-COUNT() returns the total number of values in the set.

By using the GROUP BY clause, you can calculate an aggregate value for several groups in one query. That is to say, ***aggregate functions*** collapse the individual rows and present the aggregate value for all the rows in the group.


### 1.2 Window functions

*Window functions* operate on a set of rows called a window frame. They return a single value for each row from the underlying query.

The window frame (or simply window) is defined using the OVER() clause. This clause also allows defining a window based on a specific column (similar to GROUP BY).

To calculate the returned values, window functions may use aggregate functions, but they will use them with the OVER() clause.

### 1.3 Similarities and Differences Between Window and Aggregate Functions

Now that we have seen both types of functions, we can summarize the similarities and differences between them.

***Both window functions and aggregate functions:***

- Operate on a set of values (rows).
- Can calculate aggregate amounts (e.g. AVG(), SUM(), MAX(), MIN(), or COUNT()) on the set.
- Can group or partition data on one or more columns.

***Aggregate functions with GROUP BY differ from window functions in that they:***

- Use GROUP BY() to define a set of rows for aggregation. 
- Group rows based on column values.
- Collapse rows based on the defined groups.


***Window functions differ from aggregate functions used with GROUP BY in that they:***

- Use OVER() instead of GROUP BY() to define a set of rows.
- May use many functions other than aggregates (e.g. RANK(), LAG(), or LEAD()).
- Groups rows on the row’s rank, percentile, etc. as well as its column value.
- Do not collapse rows.
- May use a sliding window frame (which depends on the current row).

## 2. Practive window functions

Let's connect the demo database, firstly.

In [2]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [3]:
%sql sqlite:///data/demo.db3

'Connected: @data/demo.db3'

### 2.1 WINDOW functions based on regular aggregate functions: AVG, MIN/MAX, SUM, COUNT

Window functions are functions that perform calculations across a set of rows related to the current row.

It is comparable to the type of calculation done with an aggregate function, but unlike regular aggregate functions, window functions do not group several rows into a single output row — the rows retain their own identities.

Behind the scenes, the window functions process more than just the query results' current row.

Let's take the ***watershed_monthly*** as the demo table.

#### 2.1.1 AVG

In [10]:
%%sql sqlite://
SELECT
    YR,
    MO,
    PREC_mm,
    ROUND(AVG(PREC_mm) OVER(PARTITION BY MO)) AS avg_PREC_mm
FROM watershed_monthly
ORDER by MO

Done.


[(1981, 1, 96.2901611328125, 114.0),
 (1982, 1, 59.744789123535156, 114.0),
 (1983, 1, 94.48113250732422, 114.0),
 (1984, 1, 111.43696594238281, 114.0),
 (1985, 1, 78.43978881835938, 114.0),
 (1986, 1, 45.433536529541016, 114.0),
 (1987, 1, 146.7594757080078, 114.0),
 (1988, 1, 106.8199234008789, 114.0),
 (1989, 1, 97.1476821899414, 114.0),
 (1990, 1, 84.20606994628906, 114.0),
 (1991, 1, 120.2723388671875, 114.0),
 (1992, 1, 114.5992660522461, 114.0),
 (1993, 1, 139.7407684326172, 114.0),
 (1994, 1, 117.82371520996094, 114.0),
 (1995, 1, 99.79764556884766, 114.0),
 (1996, 1, 171.72898864746094, 114.0),
 (1997, 1, 109.2258529663086, 114.0),
 (1998, 1, 165.39779663085938, 114.0),
 (1999, 1, 129.19338989257812, 114.0),
 (2000, 1, 81.89866638183594, 114.0),
 (2001, 1, 66.61447143554688, 114.0),
 (2002, 1, 125.24539184570312, 114.0),
 (2003, 1, 137.62448120117188, 114.0),
 (2004, 1, 116.2384262084961, 114.0),
 (2005, 1, 178.54624938964844, 114.0),
 (2006, 1, 99.33370971679688, 114.0),
 (2007, 1, 206.89495849609375, 114.0),
 (2008, 1, 77.4709701538086, 114.0),
 (2009, 1, 122.1943359375, 114.0),
 (2010, 1, 107.7326431274414, 114.0),
 (1981, 2, 160.22804260253906, 119.0),
 (1982, 2, 87.51470184326172, 119.0),
 (1983, 2, 52.12006378173828, 119.0),
 (1984, 2, 163.99578857421875, 119.0),
 (1985, 2, 130.32330322265625, 119.0),
 (1986, 2, 110.43309020996094, 119.0),
 (1987, 2, 184.14231872558594, 119.0),
 (1988, 2, 183.34051513671875, 119.0),
 (1989, 2, 117.50041961669922, 119.0),
 (1990, 2, 116.64447021484375, 119.0),
 (1991, 2, 113.96415710449219, 119.0),
 (1992, 2, 189.6724853515625, 119.0),
 (1993, 2, 136.2637939453125, 119.0),
 (1994, 2, 158.59536743164062, 119.0),
 (1995, 2, 124.73197174072266, 119.0),
 (1996, 2, 181.6288604736328, 119.0),
 (1997, 2, 110.37721252441406, 119.0),
 (1998, 2, 118.6283187866211, 119.0),
 (1999, 2, 97.8355712890625, 119.0),
 (2000, 2, 48.70695114135742, 119.0),
 (2001, 2, 47.3634033203125, 119.0),
 (2002, 2, 105.65862274169922, 119.0),
 (2003, 2, 113.89664459228516, 119.0),
 (2004, 2, 110.6496353149414, 119.0),
 (2005, 2, 56.347206115722656, 119.0),
 (2006, 2, 55.579612731933594, 119.0),
 (2007, 2, 55.45048141479492, 119.0),
 (2008, 2, 95.09587860107422, 119.0),
 (2009, 2, 161.8676300048828, 119.0),
 (2010, 2, 173.24346923828125, 119.0),
 (1981, 3, 136.6529083251953, 99.0),
 (1982, 3, 150.9921875, 99.0),
 (1983, 3, 129.8241729736328, 99.0),
 (1984, 3, 104.96453857421875, 99.0),
 (1985, 3, 99.34416198730469, 99.0),
 (1986, 3, 108.29244232177734, 99.0),
 (1987, 3, 121.24017333984375, 99.0),
 (1988, 3, 127.04043579101562, 99.0),
 (1989, 3, 115.449462890625, 99.0),
 (1990, 3, 110.31065368652344, 99.0),
 (1991, 3, 122.04442596435547, 99.0),
 (1992, 3, 160.9407501220703, 99.0),
 (1993, 3, 133.84007263183594, 99.0),
 (1994, 3, 106.6016616821289, 99.0),
 (1995, 3, 122.27830505371094, 99.0),
 (1996, 3, 132.6336212158203, 99.0),
 (1997, 3, 104.50662231445312, 99.0),
 (1998, 3, 149.83721923828125, 99.0),
 (1999, 3, 38.80699920654297, 99.0),
 (2000, 3, 19.893341064453125, 99.0),
 (2001, 3, 11.640951156616211, 99.0),
 (2002, 3, 44.062137603759766, 99.0),
 (2003, 3, 47.003868103027344, 99.0),
 (2004, 3, 49.31222152709961, 99.0),
 (2005, 3, 96.73483276367188, 99.0),
 (2006, 3, 26.39670181274414, 99.0),
 (2007, 3, 100.57988739013672, 99.0),
 (2008, 3, 52.47661590576172, 99.0),
 (2009, 3, 122.73454284667969, 99.0),
 (2010, 3, 110.05859375, 99.0),
 (1981, 4, 118.85740661621094, 69.0),
 (1982, 4, 97.7410659790039, 69.0),
 (1983, 4, 163.90794372558594, 69.0),
 (1984, 4, 95.60626220703125, 69.0),
 (1985, 4, 81.7612075805664, 69.0),
 (1986, 4, 71.4702377319336, 69.0),
 (1987, 4, 77.35852813720703, 69.0),
 (1988, 4, 82.293212890625, 69.0),
 (1989, 4, 91.74241638183594, 69.0),
 (1990, 4, 69.09127044677734, 69.0),
 (1991, 4, 105.03886413574219, 69.0),
 (1992, 4, 117.70687866210938, 69.0),
 (1993, 4, 168.93685913085938, 69.0),
 (1994, 4, 61.40770721435547, 69.0),
 (1995, 4, 85.01670837402344, 69.0),
 (1996, 4, 59.92761993408203, 6

There is no **GROUP BY** clause for the AVG function, but how does the SQL engine know which rows to use to compute the average? The answer is the **PARTITION BY** clause inside the **OVER()** utility, and we are calculating the average based on a unique value of rating.

In the final output, every row has the average PREC_mm from the same month (*MO*). By adding a ***ORDER by***, you will find there is the same averge value for each month.

#### 2.1.2 SUM

Let’s take a look at a more complicated example, where we calculated a running sum with a window function.

In [34]:
%%sql sqlite://
SELECT
    YR,
    MO,
    PREC_mm,    
    SUM(PREC_mm) OVER(ORDER BY MO) AS running_total,
    SUM(PREC_mm) OVER() AS overall,
    ROUND(SUM(PREC_mm) OVER(ORDER BY MO) * 100.0 /SUM(PREC_mm) OVER(), 2) AS running_percentage
FROM watershed_monthly

Done.


[(1981, 1, 96.2901611328125, 3408.3335914611816, 24084.3432366848, 14.15),
 (1982, 1, 59.744789123535156, 3408.3335914611816, 24084.3432366848, 14.15),
 (1983, 1, 94.48113250732422, 3408.3335914611816, 24084.3432366848, 14.15),
 (1984, 1, 111.43696594238281, 3408.3335914611816, 24084.3432366848, 14.15),
 (1985, 1, 78.43978881835938, 3408.3335914611816, 24084.3432366848, 14.15),
 (1986, 1, 45.433536529541016, 3408.3335914611816, 24084.3432366848, 14.15),
 (1987, 1, 146.7594757080078, 3408.3335914611816, 24084.3432366848, 14.15),
 (1988, 1, 106.8199234008789, 3408.3335914611816, 24084.3432366848, 14.15),
 (1989, 1, 97.1476821899414, 3408.3335914611816, 24084.3432366848, 14.15),
 (1990, 1, 84.20606994628906, 3408.3335914611816, 24084.3432366848, 14.15),
 (1991, 1, 120.2723388671875, 3408.3335914611816, 24084.3432366848, 14.15),
 (1992, 1, 114.5992660522461, 3408.3335914611816, 24084.3432366848, 14.15),
 (1993, 1, 139.7407684326172, 3408.3335914611816, 24084.3432366848, 14.15),
 (1994, 1, 117.82371520996094, 3408.3335914611816, 24084.3432366848, 14.15),
 (1995, 1, 99.79764556884766, 3408.3335914611816, 24084.3432366848, 14.15),
 (1996, 1, 171.72898864746094, 3408.3335914611816, 24084.3432366848, 14.15),
 (1997, 1, 109.2258529663086, 3408.3335914611816, 24084.3432366848, 14.15),
 (1998, 1, 165.39779663085938, 3408.3335914611816, 24084.3432366848, 14.15),
 (1999, 1, 129.19338989257812, 3408.3335914611816, 24084.3432366848, 14.15),
 (2000, 1, 81.89866638183594, 3408.3335914611816, 24084.3432366848, 14.15),
 (2001, 1, 66.61447143554688, 3408.3335914611816, 24084.3432366848, 14.15),
 (2002, 1, 125.24539184570312, 3408.3335914611816, 24084.3432366848, 14.15),
 (2003, 1, 137.62448120117188, 3408.3335914611816, 24084.3432366848, 14.15),
 (2004, 1, 116.2384262084961, 3408.3335914611816, 24084.3432366848, 14.15),
 (2005, 1, 178.54624938964844, 3408.3335914611816, 24084.3432366848, 14.15),
 (2006, 1, 99.33370971679688, 3408.3335914611816, 24084.3432366848, 14.15),
 (2007, 1, 206.89495849609375, 3408.3335914611816, 24084.3432366848, 14.15),
 (2008, 1, 77.4709701538086, 3408.3335914611816, 24084.3432366848, 14.15),
 (2009, 1, 122.1943359375, 3408.3335914611816, 24084.3432366848, 14.15),
 (2010, 1, 107.7326431274414, 3408.3335914611816, 24084.3432366848, 14.15),
 (1981, 2, 160.22804260253906, 6970.13357925415, 24084.3432366848, 28.94),
 (1982, 2, 87.51470184326172, 6970.13357925415, 24084.3432366848, 28.94),
 (1983, 2, 52.12006378173828, 6970.13357925415, 24084.3432366848, 28.94),
 (1984, 2, 163.99578857421875, 6970.13357925415, 24084.3432366848, 28.94),
 (1985, 2, 130.32330322265625, 6970.13357925415, 24084.3432366848, 28.94),
 (1986, 2, 110.43309020996094, 6970.13357925415, 24084.3432366848, 28.94),
 (1987, 2, 184.14231872558594, 6970.13357925415, 24084.3432366848, 28.94),
 (1988, 2, 183.34051513671875, 6970.13357925415, 24084.3432366848, 28.94),
 (1989, 2, 117.50041961669922, 6970.13357925415, 24084.3432366848, 28.94),
 (1990, 2, 116.64447021484375, 6970.13357925415, 24084.3432366848, 28.94),
 (1991, 2, 113.96415710449219, 6970.13357925415, 24084.3432366848, 28.94),
 (1992, 2, 189.6724853515625, 6970.13357925415, 24084.3432366848, 28.94),
 (1993, 2, 136.2637939453125, 6970.13357925415, 24084.3432366848, 28.94),
 (1994, 2, 158.59536743164062, 6970.13357925415, 24084.3432366848, 28.94),
 (1995, 2, 124.73197174072266, 6970.13357925415, 24084.3432366848, 28.94),
 (1996, 2, 181.6288604736328, 6970.13357925415, 24084.3432366848, 28.94),
 (1997, 2, 110.37721252441406, 6970.13357925415, 24084.3432366848, 28.94),
 (1998, 2, 118.6283187866211, 6970.13357925415, 24084.3432366848, 28.94),
 (1999, 2, 97.8355712890625, 6970.13357925415, 24084.3432366848, 28.94),
 (2000, 2, 48.70695114135742, 6970.13357925415, 24084.3432366848, 28.94),
 (2001, 2, 47.3634033203125, 6970.13357925415, 24084.3432366848, 28.94),
 (2002, 2, 105.65862274169922, 6970.13357925415, 24084.3432366848, 28.94),
 (2003, 2, 113.89664459228516, 6970.13357925415, 24084.3432366848, 28.94),
 

Notice that there is no **PARTITION BY** clause because I am not grouping those  into any PREC_mm sub-categories. I want to compute my overall progress but not based on any subgroups or categories.

Another thing to notice is that if I don’t add anything inside the **OVER()** function, I get the total number of PREC_mm from the entire table. As you can see from the second last column: they all have the same value of **24084.34**, but after I add the **ORDER BY** clause, I get the running total of the PREC_mm up to that specific row (**running_total** column).

#### 2.1.3 Others

You can try ***MIN***, ***COUNT**, ***AVG*** by yourself, which are all similar to prevois examples of ***SUM*** and ***MAX**.

### 2.2 ROW_NUMBER, RANK, DENSE_RANK


Let’s go through some of the essential WINDOW functions: **ROW_NUMBER** and **RANK**. In the following example, our goal is to create a ranking index based on the amount of PREC_mm for the entire table.

#### 2.2.1 ROW_NUMBER

In [14]:
%%sql sqlite://
SELECT
    YR,
    MO,
    PREC_mm,  
    ROW_NUMBER() OVER(ORDER BY PREC_mm DESC) AS row_num
FROM watershed_monthly
ORDER by row_num

Done.


[(2007, 1, 206.89495849609375, 1),
 (1992, 2, 189.6724853515625, 2),
 (2009, 12, 184.78343200683594, 3),
 (1987, 2, 184.14231872558594, 4),
 (1988, 2, 183.34051513671875, 5),
 (1996, 2, 181.6288604736328, 6),
 (2005, 1, 178.54624938964844, 7),
 (2010, 2, 173.24346923828125, 8),
 (1996, 1, 171.72898864746094, 9),
 (1991, 11, 169.23207092285156, 10),
 (1993, 4, 168.93685913085938, 11),
 (2002, 12, 165.4676055908203, 12),
 (1998, 1, 165.39779663085938, 13),
 (1984, 2, 163.99578857421875, 14),
 (1983, 4, 163.90794372558594, 15),
 (2005, 11, 162.82516479492188, 16),
 (2009, 2, 161.8676300048828, 17),
 (1992, 3, 160.9407501220703, 18),
 (1981, 2, 160.22804260253906, 19),
 (1994, 2, 158.59536743164062, 20),
 (1998, 12, 156.7340545654297, 21),
 (2010, 12, 153.69801330566406, 22),
 (2008, 12, 152.76266479492188, 23),
 (1982, 3, 150.9921875, 24),
 (1998, 3, 149.83721923828125, 25),
 (1987, 1, 146.7594757080078, 26),
 (2005, 12, 141.8644561767578, 27),
 (1993, 1, 139.7407684326172, 28),
 (2007, 10, 138.5694122314453, 29),
 (2003, 1, 137.62448120117188, 30),
 (1981, 3, 136.6529083251953, 31),
 (1993, 2, 136.2637939453125, 32),
 (1988, 11, 134.65943908691406, 33),
 (1993, 3, 133.84007263183594, 34),
 (2007, 12, 133.10231018066406, 35),
 (1996, 3, 132.6336212158203, 36),
 (1985, 2, 130.32330322265625, 37),
 (1983, 3, 129.8241729736328, 38),
 (1999, 1, 129.19338989257812, 39),
 (1988, 3, 127.04043579101562, 40),
 (2002, 1, 125.24539184570312, 41),
 (1995, 2, 124.73197174072266, 42),
 (1989, 12, 124.12555694580078, 43),
 (1993, 10, 123.36309814453125, 44),
 (2004, 10, 122.78729248046875, 45),
 (2009, 3, 122.73454284667969, 46),
 (1995, 3, 122.27830505371094, 47),
 (2009, 1, 122.1943359375, 48),
 (1991, 3, 122.04442596435547, 49),
 (1987, 3, 121.24017333984375, 50),
 (1990, 11, 120.88981628417969, 51),
 (2009, 11, 120.41217041015625, 52),
 (1991, 1, 120.2723388671875, 53),
 (1981, 4, 118.85740661621094, 54),
 (1998, 2, 118.6283187866211, 55),
 (1994, 1, 117.82371520996094, 56),
 (1992, 4, 117.70687866210938, 57),
 (1989, 2, 117.50041961669922, 58),
 (2006, 9, 117.03430938720703, 59),
 (1999, 9, 117.02091979980469, 60),
 (1990, 2, 116.64447021484375, 61),
 (2004, 1, 116.2384262084961, 62),
 (1994, 10, 116.16419219970703, 63),
 (1986, 11, 115.95671844482422, 64),
 (2008, 10, 115.79918670654297, 65),
 (1989, 3, 115.449462890625, 66),
 (1992, 1, 114.5992660522461, 67),
 (1991, 2, 113.96415710449219, 68),
 (2003, 2, 113.89664459228516, 69),
 (1982, 11, 112.13990783691406, 70),
 (1984, 1, 111.43696594238281, 71),
 (1997, 12, 110.74897003173828, 72),
 (2004, 2, 110.6496353149414, 73),
 (1986, 2, 110.43309020996094, 74),
 (2006, 12, 110.41514587402344, 75),
 (1997, 2, 110.37721252441406, 76),
 (1990, 3, 110.31065368652344, 77),
 (2010, 3, 110.05859375, 78),
 (1990, 12, 109.86883544921875, 79),
 (1987, 9, 109.45503234863281, 80),
 (1997, 1, 109.2258529663086, 81),
 (1986, 3, 108.29244232177734, 82),
 (2010, 1, 107.7326431274414, 83),
 (1982, 10, 107.47823333740234, 84),
 (1992, 11, 107.41029357910156, 85),
 (1988, 1, 106.8199234008789, 86),
 (1994, 3, 106.6016616821289, 87),
 (2002, 2, 105.65862274169922, 88),
 (2004, 11, 105.43321228027344, 89),
 (1991, 4, 105.03886413574219, 90),
 (1984, 3, 104.96453857421875, 91),
 (1997, 3, 104.50662231445312, 92),
 (2007, 3, 100.57988739013672, 93),
 (2004, 12, 99.89029693603516, 94),
 (1995, 1, 99.79764556884766, 95),
 (1985, 3, 99.34416198730469, 96),
 (2006, 1, 99.33370971679688, 97),
 (1991, 12, 98.52151489257812, 98),
 (2003, 12, 98.40425109863281, 99),
 (1999, 2, 97.8355712890625, 100),
 (1982, 4, 97.7410659790039, 101),
 (1989, 1, 97.1476821899414, 102),
 (2005, 3, 96.73483276367188, 103),
 (1993, 12, 96.52111053466797, 104),
 (1981, 1, 96.2901611328125, 105),
 (1984, 4, 95.60626220703125, 106),
 (2008, 2, 95.09587860107422, 107),
 (1983, 1, 94.48113250732422, 108),
 (2008, 11, 93.5806655883789, 109),
 (2002, 10, 93.52913665771484, 110),
 (2010, 11, 93.515625, 111),
 (2009, 9, 93.35775756835938, 112),
 (200

As you can see, the ROW_NUMBER function generates a sequence of integers, starting from 1, for each row.

#### 2.2.2 ROW_NUMBER over PARTITION

Let’s take a look at another example. Instead of comparing a amount of PREC_mm to all other PREC_mm from the entire table, we can rank them within each month using **PARTITION BY**.

In [19]:
%%sql sqlite://
SELECT
    YR,
    MO,
    PREC_mm,  
    ROW_NUMBER() OVER(PARTITION BY MO ORDER BY PREC_mm DESC) AS row_num
FROM watershed_monthly
ORDER by row_num

Done.


[(2007, 1, 206.89495849609375, 1),
 (1992, 2, 189.6724853515625, 1),
 (1992, 3, 160.9407501220703, 1),
 (1993, 4, 168.93685913085938, 1),
 (1983, 5, 88.9581527709961, 1),
 (1987, 6, 74.57366943359375, 1),
 (1983, 7, 52.85027313232422, 1),
 (2008, 8, 66.79488372802734, 1),
 (2006, 9, 117.03430938720703, 1),
 (2007, 10, 138.5694122314453, 1),
 (1991, 11, 169.23207092285156, 1),
 (2009, 12, 184.78343200683594, 1),
 (2005, 1, 178.54624938964844, 2),
 (1987, 2, 184.14231872558594, 2),
 (1982, 3, 150.9921875, 2),
 (1983, 4, 163.90794372558594, 2),
 (1982, 5, 87.78800201416016, 2),
 (1983, 6, 65.43035888671875, 2),
 (2010, 7, 40.155967712402344, 2),
 (2001, 8, 57.558448791503906, 2),
 (1999, 9, 117.02091979980469, 2),
 (1993, 10, 123.36309814453125, 2),
 (2005, 11, 162.82516479492188, 2),
 (2002, 12, 165.4676055908203, 2),
 (1996, 1, 171.72898864746094, 3),
 (1988, 2, 183.34051513671875, 3),
 (1998, 3, 149.83721923828125, 3),
 (1981, 4, 118.85740661621094, 3),
 (1981, 5, 84.54698181152344, 3),
 (1982, 6, 62.78643035888672, 3),
 (2000, 7, 36.59354782104492, 3),
 (1991, 8, 54.519954681396484, 3),
 (1987, 9, 109.45503234863281, 3),
 (2004, 10, 122.78729248046875, 3),
 (1988, 11, 134.65943908691406, 3),
 (1998, 12, 156.7340545654297, 3),
 (1998, 1, 165.39779663085938, 4),
 (1996, 2, 181.6288604736328, 4),
 (1981, 3, 136.6529083251953, 4),
 (1992, 4, 117.70687866210938, 4),
 (1989, 5, 61.8648796081543, 4),
 (1989, 6, 55.292442321777344, 4),
 (2003, 7, 35.86302185058594, 4),
 (2010, 8, 54.445960998535156, 4),
 (2009, 9, 93.35775756835938, 4),
 (1994, 10, 116.16419219970703, 4),
 (1990, 11, 120.88981628417969, 4),
 (2010, 12, 153.69801330566406, 4),
 (1987, 1, 146.7594757080078, 5),
 (2010, 2, 173.24346923828125, 5),
 (1993, 3, 133.84007263183594, 5),
 (1991, 4, 105.03886413574219, 5),
 (1991, 5, 60.08303451538086, 5),
 (1981, 6, 44.83718490600586, 5),
 (1984, 7, 34.1981315612793, 5),
 (1994, 8, 46.66091537475586, 5),
 (2004, 9, 91.2891845703125, 5),
 (2008, 10, 115.79918670654297, 5),
 (2009, 11, 120.41217041015625, 5),
 (2008, 12, 152.76266479492188, 5),
 (1993, 1, 139.7407684326172, 6),
 (1984, 2, 163.99578857421875, 6),
 (1996, 3, 132.6336212158203, 6),
 (1982, 4, 97.7410659790039, 6),
 (1993, 5, 55.35698699951172, 6),
 (1984, 6, 43.40272521972656, 6),
 (1987, 7, 33.39873504638672, 6),
 (2004, 8, 42.52296829223633, 6),
 (2000, 9, 84.15239715576172, 6),
 (1982, 10, 107.47823333740234, 6),
 (1986, 11, 115.95671844482422, 6),
 (2005, 12, 141.8644561767578, 6),
 (2003, 1, 137.62448120117188, 7),
 (2009, 2, 161.8676300048828, 7),
 (1983, 3, 129.8241729736328, 7),
 (1984, 4, 95.60626220703125, 7),
 (1985, 5, 48.69215393066406, 7),
 (1993, 6, 39.68621063232422, 7),
 (1981, 7, 32.325984954833984, 7),
 (2009, 8, 41.06700897216797, 7),
 (1993, 9, 70.33382415771484, 7),
 (2002, 10, 93.52913665771484, 7),
 (1982, 11, 112.13990783691406, 7),
 (2007, 12, 133.10231018066406, 7),
 (1999, 1, 129.19338989257812, 8),
 (1981, 2, 160.22804260253906, 8),
 (1988, 3, 127.04043579101562, 8),
 (1989, 4, 91.74241638183594, 8),
 (1988, 5, 47.07449722290039, 8),
 (1986, 6, 27.704425811767578, 8),
 (1986, 7, 28.09088134765625, 8),
 (1986, 8, 40.33757019042969, 8),
 (2003, 9, 69.0370864868164, 8),
 (1991, 10, 87.55663299560547, 8),
 (1992, 11, 107.41029357910156, 8),
 (1989, 12, 124.12555694580078, 8),
 (2002, 1, 125.24539184570312, 9),
 (1994, 2, 158.59536743164062, 9),
 (2009, 3, 122.73454284667969, 9),
 (1995, 4, 85.01670837402344, 9),
 (1998, 5, 39.28791809082031, 9),
 (1988, 6, 23.411775588989258, 9),
 (1999, 7, 27.09001350402832, 9),
 (1999, 8, 34.96633529663086, 9),
 (2005, 9, 62.30735778808594, 9),
 (2003, 10, 87.20095825195312, 9),
 (2004, 11, 105.43321228027344, 9),
 (1997, 12, 110.74897003173828, 9),
 (2009, 1, 122.1943359375, 10),
 (1993, 2, 136.2637939453125, 10),
 (1995, 3, 122.27830505371094, 10),
 (1988, 4, 82.293212890625, 10),
 (1990, 5, 39.08815002441406, 10),
 (2010, 6, 21.425451278686523, 10),
 (1989, 7, 24.988685607910156, 10),
 (2000, 8, 33.625

**ROW_NUMBER** with **OVER** and **PARTITION BY** is a regular pattern that is frequently used in advanced SQL.

#### 2.2.3 Others

Now you can try the **RANK** and ***DENSE_RANK*** functions by yourself as they are very similar to **ROW_NUMBER**. The difference between **RANK** and **ROW_NUMBER** is that **RANK** assigns the same unique values if there is a tie and restarts the next value with the total number of rows up to that row.

**DENSE_RANK**. It is very similar to **RANK** but differs in how it handles ties. It restarts with the following immediate consecutive value rather than creating a gap.

### 2.3 NTILE

**NTILE** is a handy function, especially for data analytics professionals. For example, as a data scientist, you probably need to create robust statistics such as quartile, quintile, median, etc., in your daily job, and **NTILE** makes it very easy to generate those numbers.

**NTILE** takes an argument of the number of buckets and then creates this number of buckets as equally as possible, based on how the rows are partitioned and ordered inside the **OVER** function.

In [21]:
%%sql sqlite://
SELECT
    YR,
    MO,
    PREC_mm,  
    NTILE(100) OVER(ORDER BY PREC_mm) AS percentile
FROM watershed_monthly
ORDER BY percentile;

Done.


[(2000, 6, 2.1060729026794434, 1),
 (1997, 7, 2.399183750152588, 1),
 (2003, 5, 3.178541898727417, 1),
 (1997, 6, 3.311275005340576, 1),
 (1990, 8, 3.512204170227051, 2),
 (1997, 8, 3.7355682849884033, 2),
 (1994, 6, 3.8289220333099365, 2),
 (1996, 7, 4.332550525665283, 2),
 (2001, 5, 4.57631254196167, 3),
 (2002, 6, 5.1767730712890625, 3),
 (1994, 7, 5.31728458404541, 3),
 (2006, 5, 6.052229404449463, 3),
 (2009, 6, 6.912513732910156, 4),
 (2005, 7, 7.210659503936768, 4),
 (1995, 7, 7.277742385864258, 4),
 (2007, 7, 7.517814636230469, 4),
 (2008, 5, 7.693519592285156, 5),
 (2009, 5, 7.703615188598633, 5),
 (2001, 6, 7.757592678070068, 5),
 (2004, 6, 7.882723808288574, 5),
 (1990, 7, 8.136661529541016, 6),
 (1991, 7, 9.473738670349121, 6),
 (2007, 6, 9.5333890914917, 6),
 (2007, 8, 9.534032821655273, 6),
 (2001, 7, 10.02917194366455, 7),
 (1985, 7, 10.127609252929688, 7),
 (2002, 5, 10.422792434692383, 7),
 (2003, 6, 10.518081665039062, 7),
 (2008, 4, 10.628643989562988, 8),
 (1985, 8, 10.73428726196289, 8),
 (2006, 6, 10.887991905212402, 8),
 (1998, 7, 10.958629608154297, 8),
 (1998, 8, 10.963287353515625, 9),
 (2000, 5, 11.022218704223633, 9),
 (1999, 5, 11.534664154052734, 9),
 (2006, 7, 11.554990768432617, 9),
 (2001, 3, 11.640951156616211, 10),
 (2008, 7, 11.83874797821045, 10),
 (1995, 8, 11.950968742370605, 10),
 (1992, 8, 12.080037117004395, 10),
 (2010, 9, 12.318583488464355, 11),
 (2000, 4, 12.554158210754395, 11),
 (1983, 9, 12.654004096984863, 11),
 (1995, 6, 13.049528121948242, 11),
 (2002, 7, 13.134001731872559, 12),
 (1993, 7, 13.32302474975586, 12),
 (2003, 4, 13.530596733093262, 12),
 (1991, 6, 13.56886100769043, 12),
 (2002, 8, 14.20559310913086, 13),
 (1984, 5, 14.243524551391602, 13),
 (1997, 5, 14.8255615234375, 13),
 (2008, 6, 15.11318302154541, 13),
 (1996, 8, 15.782317161560059, 14),
 (1984, 8, 15.824188232421875, 14),
 (2005, 6, 16.190933227539062, 14),
 (2001, 4, 16.332597732543945, 14),
 (1999, 10, 16.439165115356445, 15),
 (1981, 9, 16.57965660095215, 15),
 (1998, 9, 16.724767684936523, 15),
 (1985, 6, 16.78373908996582, 15),
 (2005, 5, 16.846424102783203, 16),
 (1993, 8, 16.8857421875, 16),
 (1999, 6, 16.968156814575195, 16),
 (1992, 7, 17.575836181640625, 16),
 (2004, 7, 17.656644821166992, 17),
 (1990, 10, 17.68677520751953, 17),
 (1982, 8, 17.83991050720215, 17),
 (1982, 9, 18.683820724487305, 17),
 (1992, 6, 18.79245376586914, 18),
 (1996, 6, 18.90243148803711, 18),
 (2007, 5, 19.25904083251953, 18),
 (2000, 3, 19.893341064453125, 18),
 (1990, 6, 20.2957763671875, 19),
 (1981, 8, 20.65149688720703, 19),
 (1988, 7, 20.721471786499023, 19),
 (1987, 8, 20.907390594482422, 19),
 (1998, 6, 21.41149139404297, 20),
 (2010, 6, 21.425451278686523, 20),
 (2002, 4, 21.45844268798828, 20),
 (1983, 8, 21.578163146972656, 20),
 (2009, 7, 22.063232421875, 21),
 (2004, 4, 22.2083740234375, 21),
 (1982, 7, 22.69496726989746, 21),
 (1999, 4, 22.82688331604004, 21),
 (1988, 6, 23.411775588989258, 22),
 (1992, 9, 23.477638244628906, 22),
 (1989, 8, 23.658241271972656, 22),
 (1994, 5, 24.38063621520996, 22),
 (2010, 10, 24.513809204101562, 23),
 (1989, 7, 24.988685607910156, 23),
 (2005, 8, 25.277902603149414, 23),
 (2006, 3, 26.39670181274414, 23),
 (1995, 5, 26.521377563476562, 24),
 (1996, 5, 26.67353057861328, 24),
 (2006, 4, 26.746002197265625, 24),
 (1999, 7, 27.09001350402832, 24),
 (1986, 9, 27.095539093017578, 25),
 (1996, 10, 27.233800888061523, 25),
 (1986, 6, 27.704425811767578, 25),
 (1986, 7, 28.09088134765625, 25),
 (1988, 10, 28.865161895751953, 26),
 (1991, 9, 30.19475746154785, 26),
 (2006, 8, 30.839567184448242, 26),
 (1988, 8, 30.90254783630371, 26),
 (1987, 10, 31.48656463623047, 27),
 (2003, 8, 32.29594802856445, 27),
 (1981, 7, 32.325984954833984, 27),
 (1985, 12, 32.49427032470703, 27),
 (2004, 5, 32.50706100463867, 28),
 (1998, 11, 32.96993637084961, 28),
 (1985, 10, 33.29714584350586, 28),
 (1987, 7, 33.39873504638672, 28),
 (1985, 9, 33.53099822998047, 29),
 (1995, 10, 33.60626220703125, 

You can create a few more statistics, such as **DECILES** (10 buckets) and **QUARTILES** (4 buckets).

### 2.4 LAG and Lead

**LAG** and **LEAD**, which are extremely useful for dealing with time-related data. The main difference between them is that LAG gets data from previous rows, while LEAD is the opposite, which fetches data from the following rows. We can use either one of the two functions to compare month-over-month growth. 

Their syntax is very similar to other window functions. Instead of focusing on the format of the syntax, let me show you a couple of examples.

#### 2.4.1 LAG

1. In the first step, we created a temporal table of ann_total with PREC_mm.

2. And in the second step, we appended the previous year’s Ann_PREC_mm to the current year’s using the LAG function.

3. Notice that last 2 columns of the first row are empty. It’s simply because 1981 is the first available year.

4. We also specified the offset, which is 1, so we fetch the next row. If you change this number to 2, then you compare the current year’s Ann_PREC_mm to the year before the previous year.

5. Finally, we divided the current year’s Ann_PREC_mm by the previous year’s to create our year Ann_PREC_mm growth.

In [33]:
%%sql sqlite://
WITH annul_total AS (
    SELECT
        YR,    
        SUM(PREC_mm) AS Ann_PREC_mm    
    FROM watershed_monthly
    GROUP by YR
)

SELECT
    YR,
    Ann_PREC_mm,
    LAG(Ann_PREC_mm, 1) OVER (ORDER BY YR) prev_year_total,
    ROUND(Ann_PREC_mm *1.0/LAG(Ann_PREC_mm,1) OVER (ORDER BY YR), 2) AS dod
FROM annul_total
ORDER BY YR

Done.


[(1981, 895.6051273345947, None, None),
 (1982, 884.6707439422607, 895.6051273345947, 0.99),
 (1983, 816.6605195999146, 884.6707439422607, 0.92),
 (1984, 867.5743923187256, 816.6605195999146, 1.06),
 (1985, 637.7255153656006, 867.5743923187256, 0.74),
 (1986, 733.8411979675293, 637.7255153656006, 1.15),
 (1987, 1007.8945198059082, 733.8411979675293, 1.37),
 (1988, 895.8465557098389, 1007.8945198059082, 0.89),
 (1989, 930.1055030822754, 895.8465557098389, 1.04),
 (1990, 751.4553575515747, 930.1055030822754, 0.81),
 (1991, 984.4703512191772, 751.4553575515747, 1.31),
 (1992, 907.946213722229, 984.4703512191772, 0.92),
 (1993, 1057.7733840942383, 907.946213722229, 1.17),
 (1994, 802.126204252243, 1057.7733840942383, 0.76),
 (1995, 696.8527975082397, 802.126204252243, 0.87),
 (1996, 799.967465877533, 696.8527975082397, 1.15),
 (1997, 689.3775112628937, 799.967465877533, 0.86),
 (1998, 843.4601650238037, 689.3775112628937, 1.22),
 (1999, 644.3016014099121, 843.4601650238037, 0.76),
 (2000, 497.95162630081177, 644.3016014099121, 0.77),
 (2001, 512.2509250640869, 497.95162630081177, 1.03),
 (2002, 702.0293397903442, 512.2509250640869, 1.37),
 (2003, 729.9442117214203, 702.0293397903442, 1.04),
 (2004, 818.378041267395, 729.9442117214203, 1.12),
 (2005, 855.0092720985413, 818.378041267395, 1.04),
 (2006, 612.290364742279, 855.0092720985413, 0.72),
 (2007, 822.1747102737427, 612.290364742279, 1.34),
 (2008, 740.0899133682251, 822.1747102737427, 0.9),
 (2009, 1040.9012470245361, 740.0899133682251, 1.41),
 (2010, 905.6684579849243, 1040.9012470245361, 0.87)]

#### 2.4.2 LEAD

Let’s take a look at another example. It’s very similar to the previous one, but instead of appending the previous year’s Ann_PREC_mm, we used the **LEAD** function with an offset of 1 to get the next year’s Ann_PREC_mm.

We then divided the next year’s Ann_PREC_mm by the current year’s Ann_PREC_mm to get the year-over-year growth.

Notice that last 2 columns of the last two rows are empty. It’s simply because 2010 and 2009 is the first two available years.

In [32]:
%%sql sqlite://
WITH annul_total AS (
    SELECT
        YR,    
        SUM(PREC_mm) AS Ann_PREC_mm    
    FROM watershed_monthly
    GROUP by YR
)

SELECT
    YR,
    Ann_PREC_mm,
    LEAD(Ann_PREC_mm, 2) OVER (ORDER BY YR) prev_year_total,
    ROUND(Ann_PREC_mm *1.0/LEAD(Ann_PREC_mm,2) OVER (ORDER BY YR), 2) AS dod
FROM annul_total
ORDER BY YR

Done.


[(1981, 895.6051273345947, 816.6605195999146, 1.1),
 (1982, 884.6707439422607, 867.5743923187256, 1.02),
 (1983, 816.6605195999146, 637.7255153656006, 1.28),
 (1984, 867.5743923187256, 733.8411979675293, 1.18),
 (1985, 637.7255153656006, 1007.8945198059082, 0.63),
 (1986, 733.8411979675293, 895.8465557098389, 0.82),
 (1987, 1007.8945198059082, 930.1055030822754, 1.08),
 (1988, 895.8465557098389, 751.4553575515747, 1.19),
 (1989, 930.1055030822754, 984.4703512191772, 0.94),
 (1990, 751.4553575515747, 907.946213722229, 0.83),
 (1991, 984.4703512191772, 1057.7733840942383, 0.93),
 (1992, 907.946213722229, 802.126204252243, 1.13),
 (1993, 1057.7733840942383, 696.8527975082397, 1.52),
 (1994, 802.126204252243, 799.967465877533, 1.0),
 (1995, 696.8527975082397, 689.3775112628937, 1.01),
 (1996, 799.967465877533, 843.4601650238037, 0.95),
 (1997, 689.3775112628937, 644.3016014099121, 1.07),
 (1998, 843.4601650238037, 497.95162630081177, 1.69),
 (1999, 644.3016014099121, 512.2509250640869, 1.26),
 (2000, 497.95162630081177, 702.0293397903442, 0.71),
 (2001, 512.2509250640869, 729.9442117214203, 0.7),
 (2002, 702.0293397903442, 818.378041267395, 0.86),
 (2003, 729.9442117214203, 855.0092720985413, 0.85),
 (2004, 818.378041267395, 612.290364742279, 1.34),
 (2005, 855.0092720985413, 822.1747102737427, 1.04),
 (2006, 612.290364742279, 740.0899133682251, 0.83),
 (2007, 822.1747102737427, 1040.9012470245361, 0.79),
 (2008, 740.0899133682251, 905.6684579849243, 0.82),
 (2009, 1040.9012470245361, None, None),
 (2010, 905.6684579849243, None, None)]

## Summary

Congratulations! If you have followed through all the examples and have seen most of the common **WINDOW functions/patterns**.

WINDOW functions are a family of SQL utilities, which provide some extremely powerful and useful features. But for many, since they are so foreign to standard SQL, they are difficult to learn and understand, have strange syntax—and are very often avoided. Writing a bug-free WINDOW function query could be quite challenging. It takes time and practice to become a master, and you are getting there soon.


## References and resources

https://sqlite.org/windowfunctions.html

https://learnsql.com/blog/window-functions-vs-aggregate-functions/

https://blog.jupyter.org/a-jupyter-kernel-for-sqlite-9549c5dcf551

https://blog.xojo.com/2018/12/18/sqlite-3-25-adds-window-functions-and-improves-alter-table/

https://sqlpad.io/playground/